[MRG+2] Implemented SelectFromModel meta-transformer

MechCoder opened the pull request:
Continuation of https://github.com/scikit-learn/scikit-learn/pull/3011/

Commit:
Implemented SelectFromModel meta-transformer

MechCoder commented:
@amueller @agramfort Do you agree with the general direction in which this PR is going?

Basically a meta-transformer is coupled with an estimator to provide the transform methods, instead of inheriting from `_LearntSelectorMixin`.

agramfort commented:
hum. What's the vision? too fuzzy for me now

ogrisel commented:
Can you please summarize the discussion of #3011 in the description of this PR?

MechCoder commented:
Sure, just give me some time.

jnothman commented:
#3011 was meant to pilot the idea of having selection based on coefficients/importances performed by a metaestimator rather than a mixin to predictors. Part of the point is not to clutter the individual estimators' APIs with parameters only used in rare cases. Instead, a metaestimator should:
* make code much more explicit as to the meaning of the transformation (i.e. constructing a metaestimator provides intrinsic documentation, while using a predictor as a transformer without comment is a bit awkward);
* provide a focal point for documenting what has become a fairly sophisticated `threshold` argument, and;
* potentially make it easier to play with thresholds over a pre-fitted model (although the interface for this is new territory without my magic [`freeze_model`](https://github.com/jnothman/scikit-learn/blob/remember/sklearn/memo.py#L93) wrapper, without non-standardly adding a parameter to `transform`.

By way of Zen tradeoffs, it prefers "explicit is better than implicit" over "flat is better than nested". Is the vision clearer, @agramfort?

Also, it's possible @maheshakya had no intention to continue working on this, but @MechCoder it may be wise to check before simply taking it over, even if it's stale.

jnothman commented:
The main immediate benefits of this are: there's no need to move the `threshold` parameter from `transform` to the class to make it gridsearchable; there's no need to add the mixin to all classes that have `feature_importances_` or `coef_` for the sake of consistency.

jnothman commented:
@ogrisel, I think most of the discussion on #3011 was implementation detail. A surprising amount needs to be touched to make this complete as a deprecation, including examples, tests, etc.

maheshakya commented:
Thank you @MechCoder for bringing this up again. You can continue this :)
I was not able to complete it since we did not have enough opinion whether to change every example and test according to this and so. So, as @jnothman mentioned the amount of work that needs to be done on those is quite large. I think we should wait to hear what others have to say and for a final resolution about changes.

agramfort commented:
@jnothman I think I get the design decision but my question was more the end user point of view. Somehing along the line of "Users with be able to do XXX with estimators A, B and C by only adding this piece of code in ...".

MechCoder commented:
@jnothman @maheshakya  Sorry for taking this PR over without saying. I incorrectly assumed that the lack of activity for around 6 months that the Pull Request is stale.

@agramfort A transform can meet two different things now.
1] Implementing the transform method in a transformer means transforming the sample points to a centroid space using the euclidean distances.
2] Inheriting from _LearntSelectorMixin, which means extracting the features of X which are below a certain threshold.

Though both are indeed transforming, inheriting from `_LearntSelectorMixin` classifies all models as transformers, which transform into different dimensional space. By using this metatransformer, such ambiguity is resolved.

So it goes like "Users will be able to extract the n most important features by wrapping around this Metatransformer with all estimators, without having to subclass from _LearntSelectorMixin"

agramfort commented:
ok got it. I like the idea. I guess we need to document this new
SelectFromModel class and explain its usage. It should also appear in
one of the examples.

jnothman commented:
(It's more that stale doesn't strictly imply ripe for adoption. There may
be reasons why something is stale.)

On 17 February 2015 at 04:00, Alexandre Gramfort <notifications@github.com>
wrote:

>  ok got it. I like the idea. I guess we need to document this new
> SelectFromModel class and explain its usage. It should also appear in
> one of the examples.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4242#issuecomment-74539596>
> .
>

amueller commented:
I agree with the direction, thought I didn't have too close a look. Is the idea to remove transform from the models that already have the mixin?

MechCoder committed:
fix test failures

MechCoder commented:
Sorry, for the huge delay. I just had a look at the pull request today and found out all the hard work has already been done by @maheshakya . I've updated the pull request description to list out todo's about what all is left to be done.

I will not be able to work on this in the near future. I've updated the label as easy as this should be a very good issue for a new developer.

amueller commented:
@MechCoder maybe this would be a good project for you to finish? It would be very helpful.

MechCoder commented:
OK, if you insist.

MechCoder commented:
Thinking about it, is it necessary to test each and every model that has a `coef_` and `feature_importances_` attribute after fitting. I don't think so because the implementation in SelectFromModel is independent of the implementation of the underlying base estimators.

@amueller could you verify that the tests in `test_from_model` are enough?

MechCoder commented:
Would it be useful to have a `n_features` parameter in the constructor? So that an user can select a certain number of features if he does not have an intuitive idea of the threshold before?

jnothman commented:
> Would it be useful to have a n_features parameter in the constructor? So that an user can select a certain number of features if he does not have an intuitive idea of the threshold before?

I'd say don't worry about it initially. Selection relative to threshold, vs k best vs percentile best is not very consistently dealt with across `feature_selection`, this model, and `randomized_l1` (which confusingly does not live within `feature_selection` and probably should). The latter only allows threshold, but only a number. This one allows threshold but uses magic to parse non-numeric input. `Select*` provides a different class for each selection approach. Why? Legacy. Now's not the time to fix it.

jnothman commented:
(But soon might be)

amueller commented:
@MechCoder I don't insist, if there is something else you'd rather work on, I'd be happy to discuss ;)

amueller commented on sklearn/feature_selection/from_model.py:
```diff
@@ -59,55 +147,131 @@ def transform(self, X, threshold=None):
+        self : object
+            Returns self.
+        """
+        if not (self.warm_start and hasattr(self, "estimator_")):
```
Is the ``warm_start`` parameter necessary? Calling ``fit`` on the estimator_ should always erase previous state, right?

MechCoder replied:
> Calling fit on the estimator_ should always erase previous state, right?

I'm not sure about that. This would be consistent with how the `warm_start` is used in models inheriting from `_LearntSelectorMixin` right?

amueller replied:
It should be consistent. But it looks to me that it would be without having ``warm_start`` in the ``SelectFromModel`` transformer. It seems a bit odd to me that with the parameter, you need to set ``warm_start`` in both ``SelectFromModel`` and in the model itself.

MechCoder replied:
Yes, I can just pull up `warm_start` from the `base_estimator`

amueller replied:
but why would you need to?

MechCoder replied:
For every call to fit in the metatransformer the base_estimator should be rewritten? But what if someone wants to use SelectFromModel along with warm_starting the base_estimator?

amueller replied:
but every call to fit of the base_estimator will build a new model, right?

jnothman replied:
Should not need warm_start, methinks, if we don't clone the estimator...

On 9 September 2015 at 07:55, Andreas Mueller <notifications@github.com>
wrote:

> In sklearn/feature_selection/from_model.py
> <https://github.com/scikit-learn/scikit-learn/pull/4242#discussion_r38985425>
> :
>
> > +        ----------
> > +        X : array-like of shape = [n_samples, n_features]
> > +            The training input samples.
> > +
> > +        y : array-like, shape = [n_samples]
> > +            The target values (integers that correspond to classes in
> > +            classification, real numbers in regression).
> > +
> > +        **fit_params : Other estimator specific parameters
> > +
> > +        Returns
> > +        -------
> > +        self : object
> > +            Returns self.
> > +        """
> > +        if not (self.warm_start and hasattr(self, "estimator_")):
>
> but every call to fit of the base_estimator will build a new model, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4242/files#r38985425>.
>

amueller commented:
I think the current tests are ok for the moment.

MechCoder commented:
@jnothman Thanks for the point. Yes, we can fix this up in upcoming issues.

MechCoder committed:
Remove warm start

MechCoder commented:
I removed warm_start and let the estimator do the work.

amueller commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -197,9 +197,6 @@ def check_importances(name, X, y):
         assert_equal(importances.shape[0], 10)
         assert_equal(n_important, 3)
 
-        X_new = clf.transform(X, threshold="mean")
```
I would rather catch the deprecation warning and remove the test when we remove the function.

MechCoder committed:
Catch filters instead of removing the tests

MechCoder commented:
Almost done.

I'm writing an example to couple the usage of LassoCV and SelectFromModel. Now I'm increasingly thinking to provide a threshold parameter to transform also, to prevent the multiple calls to fit?

For example, 

    clf = LassoCV()
    sfm = SelectFromModel(clf)
    sfm.fit(X, y, threshold=1.5)
    sfm.transform(X)
    # Suppose 1.5 is not enough to filter out a certain number of features, then we need to fit again with the changed thrshold/
    sfm.fit(X, y, threshold=2.5)

This can be done by providing a threshold parameter to transform.
What do you guys think? @jnothman @agramfort @amueller

amueller commented:
we have the transform in master, so I'm ok with it.
The user could always manually set the threshold before calling transform, though.

jnothman commented:
I don't understand why you're presenting `threshold` as a parameter to `fit`. I think we're trying to avoid scalar parameters to `fit` or `transform`. Keeping the `threshold` as a class parameter is consistent with `SelectKBest` etc. It would be nice if the idea that one could change it and transform without fitting again were documented.

We might also consider providing the option to memoize the base estimator fitting, but perhaps not in this version; it's a more general issue with metaestimators/pipelines.

amueller commented:
wait, yeah I was basing my comment on your discussion, not the post. It shouldn't be in ``fit``, and it might be in ``transform``. Though I would prefer it not to be in transform. Documenting the behavior sounds good.

MechCoder commented:
@jnothman Sorry, I did not mean fit but transform. I wrote the example code in haste.

MechCoder committed:
Added example to depict feature selction using SelectFromModel and Lasso

MechCoder committed:
Minor doc changes and removed _set_threshold and _set_importances

MechCoder commented:
I've added an example and changed the status to [MRG]
@jnothman Do you recall why there should be methods `_set_threshold` and `_set_importances`? I grepped but could not find them being used anywhere.

amueller commented:
Travis is unhappy (shape mismatch on old numpy?!)

jnothman commented:
> @jnothman Do you recall why there should be methods _set_threshold and _set_importances?

no.

MechCoder commented:
OK, I removed it and all tests pass.

MechCoder commented:
Unrelated test failure

MechCoder commented:
@glouppe I would appreciate it if you have time to look at this, since you were part of the initial discussion as well. It would be nice to get this in the release as this has been floating for a couple of years.

jnothman commented on doc/whats_new.rst:
```diff
@@ -184,6 +184,11 @@ Enhancements
+     be used along with estimators that have `coef_` or `feature_importances_`
+     attribute to select important features of the input data. By
+     `Maheshakya Wijewardena`_, `Joel Nothman`_ and `Manoj Kumar`_.
+
```
Also should have entry in API changes describing deprecation.

jnothman commented on examples/feature_selection/select_from_model.py:
```diff
@@ -0,0 +1,51 @@
+	sfm.threshold += 0.1
+	X_transform = sfm.transform(X)
+	n_features = X_transform.shape[1]
+
```
Should we also illustrate transform('mean')? Perhaps this belongs in narrative docs, which you are yet to commit.

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -18,6 +82,8 @@ class _LearntSelectorMixin(TransformerMixin):
     ``feature_importances_`` or ``coef_`` attribute to evaluate the relative
     importance of individual features for feature selection.
     """
+    @deprecated('Support to use estimators as feature selectors will be '
```
Could you please check that no examples or narrative doc triggers this warning? If they do, they should be changed to use `SelectFromModel`.

jnothman replied:
For example https://github.com/scikit-learn/scikit-learn/blob/0.16.0/doc/modules/feature_selection.rst#feature-selection-as-part-of-a-pipeline uses `_LearntSelectorMixin`

glouppe replied:
+1 for checking that if it is not already done

amueller replied:
+1 ^^

MechCoder replied:
checked already!

jnothman commented:
Please update the "L1-based feature selection" narrative docs at least.

maheshakya commented:
@MechCoder , this looks good to me once you attend Joel's comments.

glouppe commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +133,93 @@ def transform(self, X, threshold=None):
+    def _get_support_mask(self):
+        self.threshold_ = _calculate_threshold(self.estimator, self.scores_,
+                                               self.threshold)
+        return self.scores_ >= self.threshold_
```
I dont know how others feel about this, but I would rather have `self.scores_` computed on-the-fly instead of storing it at `fit` time. The main use case would be when you want to wrap an already-fit estimator. With the current API this is not possible. Of course, this should be combined with a check that the estimator has been/is fit.

MechCoder replied:
The idea that the `transform` in the model cannot be used without refitting the meta-transfomer has been raised by me before and a "hacky" way to overcome this has been explained in the example, i.e manually setting the `threshold` attribute.

I don't really understand what you mean. If we initialize the metatransformer with the fitted estimator, what would be the need of passing `X` or `y` to the fit method or even having a fit method? (Since the scores can be calculated at the transform method)

glouppe replied:
None. It would mean that you could transform without having to call `SelectFromModel.fit`. (But you could still do it to (re)fit the base estimator if you wanted/needed to.)

glouppe replied:
In terms of implementation it would be fairly easy to do: it boils down to replace occurences of `self.scores_` by ` _get_feature_importances(self.estimator_, X)`.

MechCoder replied:
To clarify, you mean to support both options right? Providing both the fitted and non-fitted estimator?

(Should we get this merged and do further changes in another PR, because both these changes are independent)

MechCoder replied:
Never mind. I'll do it here. @amueller is also +1 with these changes.

glouppe commented on sklearn/feature_selection/from_model.py:
```diff
@@ -1,14 +1,78 @@
+    else:
+        raise ValueError("Missing `feature_importances_` or `coef_`"
+                         " attribute, did you forget to set the "
+                         "estimator's parameter to compute it?")
```
We should update this docstring as it relates to the old `compute_feature_importances` parameter we had. I would rephrase the error by hinting that the estimator (the provided one or the meta- one) should be fit before calling `transform`.

MechCoder replied:
thanks for the catch!

glouppe commented:
Thanks a lot for these changes @MechCoder ! This is definitely a better API than what we currently have.

MechCoder committed:
Now a fitted estimator can be passed to SelectFromModel

MechCoder commented:
@glouppe I've added your suggestions in the last commit. I'll remove the deprecation warnings and add narrative docs in some time.

MechCoder committed:
Add narrative docs and fix examples

MechCoder commented:
@jnothman @glouppe Thanks for giving reviews. Mind giving final passes?

glouppe commented on sklearn/feature_selection/from_model.py:
```diff
@@ -44,59 +109,18 @@ def transform(self, X, threshold=None):
-            else:
-                importances = np.sum(np.abs(self.coef_), axis=0)
-
+        importances = _get_feature_importances(self)
```
Is this correct? Shouldn't it be called on `self.estimator` or `self.estimator_`?

glouppe replied:
Oh sorry never mind, this is the deprecated class.

glouppe commented on sklearn/utils/testing.py:
```diff
@@ -528,8 +528,8 @@ def uninstall_mldata_mock():
-OTHER = ["Pipeline", "FeatureUnion", "GridSearchCV",
-         "RandomizedSearchCV"]
+OTHER = ["Pipeline", "FeatureUnion", "GridSearchCV", "RandomizedSearchCV",
+         "GridSearchCV", "SelectFromModel"]
```
Dont know why but `GridSearchCV` is now twice in the list.

glouppe commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -41,3 +48,83 @@ def test_invalid_input():
+
+    clf.fit(iris.data, iris.target)
+    model = SelectFromModel(clf)
+    assert_array_equal(model.transform(iris.data), X_transform)
```
I think you should add a few tests for how `threshold` can be specified.

jnothman replied:
I'm not sure exactly what @glouppe is referring to, whether he means strings, etc. But I'd like to see a test that what you do in your Boston example will work: modifying the threshold without re-fitting.

glouppe commented on examples/feature_selection/plot_select_from_model_boston.py:
```diff
@@ -0,0 +1,51 @@
+while n_features > 2:
+	sfm.threshold += 0.1
+	X_transform = sfm.transform(X)
+	n_features = X_transform.shape[1]
```
Use spaces instead of tabs

glouppe commented on examples/feature_selection/plot_select_from_model_boston.py:
```diff
@@ -0,0 +1,51 @@
+# Plot the selected two features from X.
+plt.title(
+	"Features selected from Boston using SelectFromModel with "
+	"threshold %0.3f." % sfm.threshold)
```
same here

glouppe commented:
A few nitpicks left, but otherwise +1 for merge once they have been fixed. Thanks for the work @MechCoder :)

MechCoder commented:
fixed your comments and updated the PR to MRG + 1

jnothman commented on doc/modules/feature_selection.rst:
```diff
@@ -131,6 +131,26 @@ number of features.
+estimator that has a ``coef_`` or ``feature_importances_`` attribute after fitting.
+It should be given a threshold parameter below which the features are considered
+unimportant and removed. If one has no idea of the prior value of the threshold,
+string inputs like "mean" or "median" or even values like "0.1*mean" can be given
```
a few more ``s here would be ideal.

jnothman commented on doc/modules/feature_selection.rst:
```diff
@@ -131,6 +131,26 @@ number of features.
+      threshold beforehand.
+
 
 .. _l1_feature_selection:
```
You should probably just merge the "L1-based feature selection" and "Selecting non-zero coefficients" as a subheading of "Feature selection using SelectFromModel"

jnothman commented on doc/modules/feature_selection.rst:
```diff
@@ -131,6 +131,26 @@ number of features.
 
+.. _select_from_model:
+
+Feature selection using SelectFromModel
```
Perhaps just "Feature selection from estimated models" or something...

MechCoder replied:
I think it might be more useful to explicitly mention the name of the model in the title.

jnothman commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -41,3 +52,96 @@ def test_invalid_input():
+
+def test_fitted_estimator():
+    """Test that a fitted estimator can be passed to SelectFromModel."""
+    clf = SGDClassifier(alpha=0.1, n_iter=10, shuffle=True, random_state=0)
```
Perhaps note "no fit required"...

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +135,99 @@ def transform(self, X, threshold=None):
+    estimator : object
+        The base estimator from which the transformer is built.
+        This can be both a fitted or a non-fitted estimator.
+        If it a fitted estimator, then transform can be called directly,
```
Please at least use `` around method names.

MechCoder committed:
Merge SelectFromModel and L1-selection examples
Add test to check the threshold can be set without refitting

MechCoder commented:
@jnothman I think I've addressed your comments.

MechCoder commented:
@jnothman I understand that you are busy but would be great if you could give some final comments.

jnothman commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -194,15 +196,18 @@ def test_probability():
-    X_new = est.transform(X, threshold="mean")
-    assert_less(X_new.shape[1], X.shape[1])
+    clean_warning_registry()
+    with warnings.catch_warnings(record=True) as record:
```
Could you please use `assert_warns` here? Apart from anything else, when deprecation concludes, the test will fail until the assertion is removed.

jnothman replied:
The second half of that comment was irrelevant. Still, `assert_warns` would be cleaner

jnothman commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -295,12 +297,16 @@ def test_feature_importances():
         clf.fit(X, y)
         assert_true(hasattr(clf, 'feature_importances_'))
+        clean_warning_registry()
+        with warnings.catch_warnings(record=True) as record:
```
"

jnothman commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -295,12 +297,16 @@ def test_feature_importances():
 
-        X_new = clf.transform(X, threshold="mean")
-        assert_less(X_new.shape[1], X.shape[1])
+            X_new = clf.transform(X, threshold="mean")
```
"

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +135,98 @@ def transform(self, X, threshold=None):
+        self.threshold = threshold
+
+    def _get_support_mask(self):
+        # SelectFromModel can directly call on transform.
```
Did we really decide this? I get why this might be helpful, but it breaks all sorts of conventions, e.g. if estimator is not fitted we should be raising `NotFittedError`. Note `CalibratedClassifierCV` supports pre-fitted estimators, but only with an explicit parameter `cv='prefit'`.

I would rather we have a standard way to freeze estimators for inclusion in various contexts. I have proposed such a helper class, but it and some code that supports it (#5080) doesn't seem to garner much interest.

In this case I think that this class is really not very interesting except for its use in a pipeline or similar. If one wants an ad-hoc mask over the `feature_importances_` attribute, it's not especially hard to get right.

MechCoder replied:
This was suggested by @glouppe here (https://github.com/scikit-learn/scikit-learn/pull/4242#discussion-diff-40066983). I do not really have an opinion on this. But I do agree that it violates the `fit-predict` API and it is almost trivial to mask the feature importances oneself.

But I suppose we can leave this debatable feature open as an issue and remove it for now?

MechCoder replied:
Or ofcourse we could have an explicit parameter that does it.

jnothman commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -24,7 +33,9 @@ def test_transform_linear_model():
                 clf.fit(X, iris.target)
-                X_new = clf.transform(X, thresh)
+                clean_warning_registry()
+                with warnings.catch_warnings(record=True) as record:
```
assert_warns.

I note that all these tests will disappear with deprecation... a little comment to that effect might be helpful for maintenance.

MechCoder replied:
related issue https://github.com/scikit-learn/scikit-learn/issues/5343

jnothman replied:
well yes, though these tests will by-and-large be deleted.

jnothman commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -377,9 +380,12 @@ def test_importances():
-        assert_less(X_new.shape[1], X.shape[1], "Failed with {0}".format(name))
+
+        clean_warning_registry()
+        with warnings.catch_warnings(record=True) as record:
```
assert_warns

jnothman commented:
[I'm undecided about the prefit support](https://github.com/scikit-learn/scikit-learn/pull/4242#discussion_r41452559), and would appreciate the opinion of @amueller or similar. But other than that and my nitpicks in testing, this looks good to me.

amueller commented:
I'd be fine with either adding a parameter ``prefit=False`` that needs to be explicitly set or removing the support.

MechCoder committed:
1. Added parameter prefit to pass in a fitted estimator.
2. Use assert_warns instead of catch_warnings
3. Remove depracation warnings in common tests.

MechCoder commented:
@jnothman @amueller  I've addressed your comments. Would be great if you could give final passes.

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +139,115 @@ def transform(self, X, threshold=None):
+        "mean" is used by default.
+
+    prefit: bool, default True
+        Whether a prefit model can be passed or not.
```
"can be passed" -> "is expected". Add "If True, this SelectFromModel cannot be used with `cross_val_score`, `GridSearchCV` and similar utilities that clone the estimator."

Because of such caveats, I think this is all quite yuck.

MechCoder replied:
But why would someone use `cross_val_score` with a fit model?

jnothman replied:
To work out which feature selection threshold is most effective!

jnothman replied:
Perhaps that answered for `GridSearchCV`...

MechCoder replied:
Actually my initial question was different, but now I really don't understand what you mean. (It might be just because I'm a bit sleepy).

The `threshold` parameter is not used at fit time (only at transform time) so how would it be able to do this selection?

jnothman replied:
Consider `GridSearchCV(make_pipeline(SelectFromModel(prefit_model, prefit=True), SVC()), {'selectfrommodel__threshold': np.linspace(0, 1)})`. This will not work due to cloning.

MechCoder replied:
Ah you meant in a pipeline. Sorry that was not obvious to me.

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +139,115 @@ def transform(self, X, threshold=None):
+        available, the object attribute ``threshold`` is used. Otherwise,
+        "mean" is used by default.
+
+    prefit: bool, default True
```
space before `:`

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +139,115 @@ def transform(self, X, threshold=None):
+        The base estimator from which the transformer is built.
+        This can be both a fitted (if ``prefit`` is set to True)
+        or a non-fitted estimator.
+        If it a fitted estimator, then ``transform`` can be called directly,
```
can -> must. Fitting isn't an option.

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +139,115 @@ def transform(self, X, threshold=None):
+    ----------
+    estimator : object
+        The base estimator from which the transformer is built.
+        This can be both a fitted (if ``prefit`` is set to True)
```
I suspect this should all be described under the `prefit` parameter description.

MechCoder commented:
@jnothman Addressed your comments. Anything else?

amueller commented on doc/modules/feature_selection.rst:
```diff
@@ -131,33 +131,50 @@ number of features.
+
+:class:`SelectFromModel` is a meta-transformer that can be used along with any
+estimator that has a ``coef_`` or ``feature_importances_`` attribute after fitting.
+It should be given a ``threshold`` parameter below which the features are considered
```
maybe remove "below"?

amueller replied:
maybe remove "below"?

amueller commented on doc/modules/feature_selection.rst:
```diff
@@ -131,33 +131,50 @@ number of features.
+:class:`SelectFromModel` is a meta-transformer that can be used along with any
+estimator that has a ``coef_`` or ``feature_importances_`` attribute after fitting.
+It should be given a ``threshold`` parameter below which the features are considered
+unimportant and removed. If one has no idea of the prior value of the threshold,
```
Apart from specifying the threshold numerically, there are build-in heuristics for finding a threshold, which can be selected using a string argument. Available heuristics are ``"mean``", ``"median"`` and  float multiples of these like ``"0.1*mean"``. [maybe something along these lines?]

amueller commented on doc/whats_new.rst:
```diff
@@ -256,7 +261,7 @@ API changes summary
       as features or as samples. All data arrays are now expected
       to be explicitly shaped ``(n_samples, n_features)``.
-      By `Vighnesh Birodkar`_.
+      By `Vighnesh Birodkar`_
```
why that change?

amueller commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +139,117 @@ def transform(self, X, threshold=None):
+        This can be both a fitted (if ``prefit`` is set to True)
+        or a non-fitted estimator.
+
+    threshold : string, float, optional
```
document default=None?

amueller commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +139,117 @@ def transform(self, X, threshold=None):
+        importance is greater or equal are kept while the others are
+        discarded. If "median" (resp. "mean"), then the ``threshold`` value is
+        the median (resp. the mean) of the feature importances. A scaling
+        factor (e.g., "1.25*mean") may also be used. If None and if
```
I don't see where an object attribute "threshold" is used. Also you should document the L1 special case.

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
+        assert_raises(ValueError, model.transform, iris.data)
+
+
+def test_validate_estimator():
```
can you add a comment what this test does?

MechCoder replied:
I did not add this, but I think the point is that the input estimator is unchanged?

amueller replied:
yeah I guess so. Can you add a comment? and actually you could use "is". Not sure what equality of estimators is.

amueller replied:
actually maybe remove the test? Or I don't understand it.

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
+
+
+def test_feature_importances():
+    X, y = datasets.make_classification(n_samples=1000,
```
can you reformat to use less lines?

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
+    transformer = SelectFromModel(estimator=est)
+    transformer.fit(X, y, sample_weight=sample_weight)
+    importances = transformer.estimator_.feature_importances_
+    assert_less(importances[1], X.shape[1])
```
What does this test? And isn't that testing just something in the forest unspecific to the SelectFromModel?

MechCoder replied:
Ouch. I did not write this as well :P

amueller replied:
where are they from? #3011 ?

MechCoder replied:
yes, the first commit. I should have reviewed it thoroughly

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
+    assert_almost_equal(importances, importances_bis)
+
+
+def test_partial_fit():
```
can you also check that ``transform`` works?

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
+    transformer.partial_fit(iris.data, iris.target,
+                            classes=np.unique(iris.target))
+    id_2 = id(transformer.estimator_)
+    assert_equal(id_1, id_2)
```
is this more elegant than just using ``assert_true(old_model is new_model)``?

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
-    assert_raises(ValueError, clf.transform, iris.data, "gobbledigook")
-    assert_raises(ValueError, clf.transform, iris.data, ".5 * gobbledigook")
+    model = SelectFromModel(clf, prefit=True)
+    assert_array_equal(model.transform(iris.data), X_transform)
```
Can you check that fit is called again even if the model is fitted if ``prefit=False``?
There are four cases we need to check the right behavior for, ``prefit=True`` and ``prefit=False`` in combination with a fitted and not fitted model, right?

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
+    X_new = transformer.transform(X)
+    assert_less(X_new.shape[1], X.shape[1])
+
+    feature_mask = (transformer.estimator_.feature_importances_ >
```
shouldn't there be a manual check like this for the different settings of ``threshold``? I think the current tests would all pass if ``threshold`` would always default to ``mean``, right?

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -37,7 +46,115 @@ def test_transform_linear_model():
+
+    # Calculate the threshold from the estimator directly.
+    est.fit(iris.data, iris.target)
+    threshold = 0.5 * np.mean(est.feature_importances_)
```
can you use est directly? if ``threshold`` is just ignored this test would pass. Or check that passing a float works somewhere else.

amueller commented:
doctests raised some ``Function transform is deprecated`` warnings.

amueller commented:
``sklearn.utils.tests.test_estimator_checks.test_check_estimator`` also has one. the rest seems good. [you could raise instead to see if it happens in the tests]

MechCoder committed:
Refactor tests

MechCoder commented:
@amueller thanks for your comments. The last commit should address all of them.

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -111,3 +141,118 @@ def transform(self, X, threshold=None):
+        This can be both a fitted (if ``prefit`` is set to True)
+        or a non-fitted estimator.
+
+    threshold : string, float, optional default None
```
I think this is one of those cases where `default None` is not very helpful, but the description should say "By default, [this is its behaviour]..."

MechCoder replied:
I have mentioned that below.

jnothman replied:
What I meant is that "default None" could be dropped, but say "By default,..." rather than "When None,..." below. But it's a really minor nitpick.

jnothman commented:
Once those are fixed up, it's good for merge.

MechCoder commented:
@jnothman I'll merge after Travis passes?

jnothman commented:
Yes, with two whats_new entries: one under API changes to tell people their `transform`s will be disappearing, and one under new features. thank you.

MechCoder commented:
Thanks a lot! @maheshakya @glouppe @amueller @jnothman

amueller commented:
:beers: was lange währt währt endlich gut. Thanks @MechCoder :)
